### PR TITLE
refact(upgrade): added switch case for 1.2 upgrade and command for upgradetask

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**What this PR does / why we need it**:
This PR adds the switch case required to upgrade from OpenEBS 1.1 to 1.2. 
It also adds a new command to read inputs from `UpgradeTask` CR instead of flags.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: cmd/upgrade/executor/cstor_spc.go [hunk 1]
```diff
@@ -103,6 +103,18 @@ func (u *UpgradeOptions) RunCStorSPCUpgrade(cmd *cobra.Command) error {
 			fmt.Println(err)
 			return errors.Errorf("Failed to upgrade cStor SPC %v:", u.cstorSPC.spcName)
 		}
+	case "1.1.0-1.2.0":
```
Comment (kmova) [inline cmd/upgrade/executor/cstor_spc.go]: Can we use a multiple case statement like: 
```
case "1.0.0-1.1.0", "1.0.0-1.2.0", "1.1.0-1.2.0":
```

Example: https://github.com/golang/go/wiki/Switch

File: cmd/upgrade/executor/cstor_spc.go [hunk 2]
```diff
@@ -103,6 +103,18 @@ func (u *UpgradeOptions) RunCStorSPCUpgrade(cmd *cobra.Command) error {
...
+		fmt.Println("Upgrading to 1.2.0")
+		err := upgrade100to110.Exec(u.fromVersion, u.toVersion,
```
Comment (kmova) [inline cmd/upgrade/executor/cstor_spc.go]: Also, should we refactor this to keep it generic as we are upgrading multiple cases using same package. 

```
upgrade100to110 "github.com/openebs/maya/pkg/upgrade/1.0.0-1.1.0/v1alpha1"
```

File: cmd/upgrade/executor/upgradetask.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+// UpgradeTaskOptions stores information required for upgradeTask upgrade
+type UpgradeTaskOptions struct {
+	resourceName string
+}
+
+// NewUpgradeTaskJob upgrade a resource from upgradeTask
+func NewUpgradeTaskJob() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "upgrade-task",
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: change: `upgrade-task` to `resource`

File: cmd/upgrade/executor/upgradetask.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		Short:   "Upgrade UpgradeTask Resource",
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: `Upgrade a resource using the details specified in the UpgradeTask CR.`

File: cmd/upgrade/executor/upgradetask.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+	errors "github.com/openebs/maya/pkg/errors/v1alpha1"
+	upgrade100to110 "github.com/openebs/maya/pkg/upgrade/1.0.0-1.1.0/v1alpha1"
+	utask "github.com/openebs/maya/pkg/upgrade/v1alpha2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	upgradeTaskJobUpgradeCmdHelpText = `
+This command upgrades the resource mentioned in upgradeTask env
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: This command upgrades the resource mentioned in the UpgradeTask CR. 
The name of the UpgradeTask CR is extracted from the ENV `UPGRADE_TASK`

File: cmd/upgrade/executor/cstor_spc.go (hunk 2 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/cstor_spc.go]: Should it be `upgrade100to120 "github.com/openebs/maya/pkg/upgrade/1.0.0-1.1.0/v1alpha1"`
Will change the directory to `github.com/openebs/maya/pkg/upgrade/1.0.0-1.2.0` in next PR

File: cmd/upgrade/executor/cstor_spc.go (hunk 1 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/cstor_spc.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 3 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 4 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 5 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/setup_job.go [hunk 6]
```diff
@@ -42,6 +42,7 @@ func NewJob() *cobra.Command {
 		NewUpgradeJivaVolumeJob(),
 		NewUpgradeCStorSPCJob(),
 		NewUpgradeCStorVolumeJob(),
+		NewUpgradeTaskJob(),
```
Comment (kmova) [inline cmd/upgrade/executor/setup_job.go]: `NewUpgradeTaskJob` -> `NewUpgradeResource`

File: cmd/upgrade/executor/upgradetask.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
+/*
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: change this to "resource.go"

File: cmd/upgrade/executor/upgradetask.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/spf13/cobra"
+
+	errors "github.com/openebs/maya/pkg/errors/v1alpha1"
+	upgrade100to120 "github.com/openebs/maya/pkg/upgrade/1.0.0-1.1.0/v1alpha1"
+	utask "github.com/openebs/maya/pkg/upgrade/v1alpha2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	upgradeTaskJobUpgradeCmdHelpText = `
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: `upgradeTaskJobUpgradeCmdHelpText` -> `resourceUpgradeCmdHelpText`

Let us rename the variables with `resource` to maintain the readability and avoid confusing with UpgradeTask CR as well as UpgradeTaskJob etc.

File: cmd/upgrade/executor/upgradetask.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+This command upgrades the resource mentioned in the UpgradeTask CR.
+The name of the UpgradeTask CR is extracted from the ENV UPGRADE_TASK
+
+Usage: upgrade resource
+`
+)
+
+// UpgradeTaskOptions stores information required for upgradeTask upgrade
+type UpgradeTaskOptions struct {
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: UpgradeResourceOptions or ResourceOptions

File: cmd/upgrade/executor/upgradetask.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+			util.CheckErr(options.InitializeDefaults(cmd), util.Fatal)
+			util.CheckErr(options.RunUpgradeTaskUpgrade(cmd), util.Fatal)
+		},
+	}
+
+	return cmd
+}
+
+// InitializeFromUpgradeTask will populate the UpgradeOptions from given UpgradeTask
+func (u *UpgradeOptions) InitializeFromUpgradeTask(cmd *cobra.Command) error {
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: InitializeFromUpgradeTaskResource

File: cmd/upgrade/executor/upgradetask.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	utaskName := getUpgradeTask()
```
Comment (kmova) [inline cmd/upgrade/executor/upgradetask.go]: upgradeTaskCRName

File: cmd/upgrade/executor/setup_job.go (hunk 6 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/setup_job.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 7 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 8 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 9 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 10 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/upgradetask.go (hunk 11 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/upgradetask.go]: Done

File: cmd/upgrade/executor/resource.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	case upgradeTaskCRObj.Spec.ResourceSpec.CStorVolume != nil:
+		u.resourceKind = "cstorVolume"
+		u.resource.name = upgradeTaskCRObj.Spec.ResourceSpec.CStorVolume.PVName
+	}
+
+	return nil
+}
+
+// RunResourcekUpgradeChecks will ensure the sanity of the upgradeTask upgrade options
+func (u *UpgradeOptions) RunResourcekUpgradeChecks(cmd *cobra.Command) error {
```
Comment (vishnuitta) [inline cmd/upgrade/executor/resource.go]: spelling mistake

Comment (shubham14bajpai) [reply cmd/upgrade/executor/resource.go]: Done

**Pull request merged**